### ISSUE TITLE
All Predictions have same value: possible debug hints

### ISSUE THREAD

**Issue by simpsus** [jefferythewind/warpgbm#7]
I know that the best practice is to provide a minimal working example of the issue. I will check if the issue persists witgh random data, if so, then I will provide the lines. For the time being, I would like to discuss possible debug options, maybe with ideas of including optional debug output in future versions.

I am on windows, python 3.11, latest warpgbm from github. 
The model construction and fitting is faily simple:

`model = WarpGBM(max_depth=6, n_estimators=20000, learning_rate=0.001, colsample_bytree=0.1)
era_id = EraSplitColumn().all(data)
model.fit(
    X=X.values,
    y=y.values,
    era_id=era_id
)`

whereas X and y are parts of a pandas Datafream and era_id is a numpy array.
The length of the data is faily large
`X.shape = (5766939, 2376)`
and in era_id there are more than 1k unique eras.

First thing that makes me wonder is fitting speed. The input is integer pre-binned and warpgbm is fast, but I find not even 5 minutes for 20k trees suspiciously fast.

inspecting the fitted model looks fine:

![Image](https://github.com/user-attachments/assets/46209726-b933-44aa-beff-8e8bac60f3b8)

Is all what I would expect. First suspicious thing is the `per_era_gain`

![Image](https://github.com/user-attachments/assets/cb5e98b8-78ad-468b-a947-fac086026cec)

So there is only zeros in the per-era gain.

![Image](https://github.com/user-attachments/assets/5f57e9f0-0b6a-4d7d-b44c-071ad39a0125)

![Image](https://github.com/user-attachments/assets/45e9dc8f-ba5b-4e40-8104-5306185c21f8)

Same goes for the training loss and the root gradient histogram.

And here I am a little stuck. What could be debug steps to narrow down the issue?

**Comment by jefferythewind** [comment]
Thanks for help here. One thing i just noticed is that training loss is only updated if you provide a validation set, updated at the intervals defined using those parameters. That explains why you don't have training loss data. perhaps that could be done better.

Also since you're using so many eras, most likely the trees are stopping growth pretty early since each split needs min_child_weight number of samples on each side of the split in each era to be a valid split.

First thing I might try is using a validation set, that will show you training and val stats during training. 

Also, check model.forest, which is the entire forest structure that is learnt from the data. That will show you how many trees you have and the structure of each tree. 

Also besides more transparency, what is the problem with this model, does it produce crap results?

**Comment by simpsus** [comment]
each era has ~5k samples. so there should be min_child_weight samples available.
Looking at `model.forest` gives insights:

```
[{'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939},
...
 {'leaf_value': 3.016334959582423e-09, 'samples': 5766939}]
```

so it produces the same tree 20k times. And each tree has only one leaf with all the samples. That explains both the fast fitting time and the prediction behaviour: The predictions are all the same, no matter which features (in sample or out of sample) I submit.

I will investigate the era sizes and ids if I can get it to change the behaviour.
I will include an eval set to get more information on the fitting process.
Besides: such a forest is never going to be useful. We could think about detecting this fitting behaviour and alerting the user that something is wrong.

**Comment by simpsus** [comment]
Update. Comments/hints always welcome.

The issue does not seem to be related to the number of eras. My first idea was that it is related to the numbering of the eras, because my first attempt did not start with 1. That is not the issue. If I start the era numbering at 1, the issue still persists. 

Further investigation: is there an era with little/no samples?

**Comment by jefferythewind** [comment]
good question. your forest structure implies that growth is terminating at the root node for some reason. In the root node all predictions get the same value. So what are possible reasons for growth terminating at the root node?

actually there are only a few options, all of which appear inside the grow_tree function.

**Comment by jefferythewind** [comment]
can check lines 358 to 361 of core.py, basically there are not any valid splits? It still could be a bug or maybe we need to slightly change how the split mask is computed? but a valid split must be the maximum directional agreement split and also have an average split gain greater than min_split_gain. I guess when there are so many eras, the best directional split may always results in a mean split gain less than the threshold (not really surprising), resulting in no valid splits.

**Comment by simpsus** [comment]
`        if not dir_score_mask.any():
            return -1, -1`

is the termination of the find_best_split method. Before that, the directional_agreement and era_splitting_criterion are both all zeros. I will investigate further

**Comment by Pranshu-Bahadur** [comment]
> Update. Comments/hints always welcome.
> 
> The issue does not seem to be related to the number of eras. My first idea was that it is related to the numbering of the eras, because my first attempt did not start with 1. That is not the issue. If I start the era numbering at 1, the issue still persists.
> 
> Further investigation: is there an era with little/no samples?

Ok I will look into this too, I have a hunch that its no samples. Unless you've concluded otherwise? Btw we have another implementation of DES in the des_branch - just fyi if you want to experiment more

**Comment by simpsus** [comment]
When the tree creation fails, in the beginning of `grow_tree` the argument histograms are both complete zeros. If the tree construction works, not.
So ... To me it seems like the problem is **before** the grow_tree method

The grow_tree method initializes them with zeroes, so I assume the `node_kernel.compute_histogram3` does not change them.
I can however not look into that method.
The minimum number of samples in an era is >3k.
There are two other histogram kernels available. The readme.md says there is a parameter. I think histogram3 is currently hardcoded. Moreso, the binding is created in the kernel and no other binding is created. So I think I can neither understand what is going on or switch to another histogram kernel.

**Comment by jefferythewind** [comment]
yeah we need to update the readme, the other 2 histogram kernels are deprecated. There is no parameter for that anymore.

**Comment by jefferythewind** [comment]
I think i might know what is going on. I believe the problem could be shared memory overflow in the histogram builder. One thing to try is to set num_bins == 5, since we don't have a method to set that automatically if your num bins implied by the data (5) is less then what you set as a hyper parameter. I will fix this. If your num bins is 10 (default) and the max shared memory per block is 48kb then we will overflow that amount and cause a silent death of the cuda launcher. So try explicitly setting num_bins to 5 or 3 and see if the error persists. In the mean time I'm marking this as a bug and will fix the implicit bin problem asap. Not sure if it will clear up your issue but it is a bug none the less.

**Comment by Pranshu-Bahadur** [comment]
Hey @simpsus have added a new histogram kernel in this branch:
[des_era_tiling_pranshu](https://github.com/jefferythewind/warpgbm/tree/des_era_tiling_pranshu)
it should work with 1000+ eras and 10+ bins usecase too now.
Thank you for all your help, let me know if there are still any bugs with the new branch 

You can do this by cloning the repo and installing on local like

```bash
git clone -b des_era_tiling_pranshu https://github.com/jefferythewind/warpgbm.git
cd warpgbm
pip install -e .  --no-build-isolation
```

**Comment by simpsus** [comment]
creating the model with `num_bins=5` fixes the issue of histogram creation. currently it is running. looks fine execpt all trees get size 4. with full numerai data it is quite slow. I will inspect the forest once a real forest is there

**Comment by simpsus** [comment]
> > Update. Comments/hints always welcome.
> > The issue does not seem to be related to the number of eras. My first idea was that it is related to the numbering of the eras, because my first attempt did not start with 1. That is not the issue. If I start the era numbering at 1, the issue still persists.
> > Further investigation: is there an era with little/no samples?
> 
> Ok I will look into this too, I have a hunch that its no samples. Unless you've concluded otherwise? Btw we have another implementation of DES in the des_branch - just fyi if you want to experiment more

Hi @Pranshu-Bahadur, 
I checked out your branch, as I am interested in the new histogram kernel. However I am hit with the installation issue again. The fix last time (renaming the .pyd file removing the platform info) is not possible, becuase there is no .pyd file created during install. Any idea what could be the cause of this?

**Comment by Pranshu-Bahadur** [comment]
> > > Update. Comments/hints always welcome.
> > > The issue does not seem to be related to the number of eras. My first idea was that it is related to the numbering of the eras, because my first attempt did not start with 1. That is not the issue. If I start the era numbering at 1, the issue still persists.
> > > Further investigation: is there an era with little/no samples?
> > 
> > Ok I will look into this too, I have a hunch that its no samples. Unless you've concluded otherwise? Btw we have another implementation of DES in the des_branch - just fyi if you want to experiment more
> 
> Hi @Pranshu-Bahadur, 
> I checked out your branch, as I am interested in the new histogram kernel. However I am hit with the installation issue again. The fix last time (renaming the .pyd file removing the platform info) is not possible, becuase there is no .pyd file created during install. Any idea what could be the cause of this?

@simpsus 
Can you try these so that we can see what's going wrong in the log output?
```shell
git clone -b des_era_tiling_pranshu https://github.com/jefferythewind/warpgbm.git
cd warpgbm                   
pip uninstall -y warpgbm
pip install -v -e . --no-build-isolation 2>&1 | tee build.log
```

**Comment by simpsus** [comment]
[build.log](https://github.com/user-attachments/files/20985922/build.log)

Now it's clear what's going on here. CUDA is not available. Why, I have no idea, but that gives me something to work with

**Comment by simpsus** [comment]
[build.log](https://github.com/user-attachments/files/20986118/build.log)

okay, uninstalled torch, reinstalled and then warpgbm actually built the extensions

**Comment by Pranshu-Bahadur** [comment]
> [build.log](https://github.com/user-attachments/files/20986118/build.log)
> 
> okay, uninstalled torch, reinstalled and then warpgbm actually built the extensions

@simpsus That's great to hear! Were you able to try our new kernel?

Lmk what you think - any feedback is always welcome


PS: What's torch version & gpu(s) are you using? (If you don't mind telling)
